[StructuralMechanicsApplication] Rve periodic

RiccardoRossi opened the pull request:
This branch provides the tools needed (an utility and an AnalysisStage) for the evaluation of the Elastic matrix corresponding to a given RVE.

The limiation is that the geoemtry of the RVE must be periodic, and it must be delimited by a parallelepiped

RiccardoRossi committed:
adding minimal classes for application of periodic boundary conditions on rve

RiccardoRossi committed:
polishing and removing debug output

RiccardoRossi committed:
working version with periodic BCs

philbucher commented:
wow this is quite messy ... you sure you want to commit like this?
Asking first before I review in detail

RiccardoRossi committed:
e cosmetic polishing + a generality improvement (possibility of prescribing the periodic variable)

Signed-off-by: riccardo <rrossi@cimne.upc.edu>

RiccardoRossi committed:
removing one unneeded solve step

RiccardoRossi committed:
adding doxygen comments

RiccardoRossi committed:
applying the protected naming convention to some of the functions in the analysisÇ

RiccardoRossi commented:
Hi @philbucher i cleaned up some (purely cosmetic) but from my point of view this is not too bad :-D

i have two main things to discuss which is the reason of being of this PR

1 - For the input, i am not sure of wherein the settings i should put the parameters to be passed to my analysis
2 - in order to have a test, i need my analysis stage to return something (the estimated matrix). We did not decide yet how this should be done. I could write to a file but it is not too clean, so i think this is something for discussion (however this PR provides a nice feature which i would love to have in for release, so if the good solution of this requires a lot of changes, i would do that in a second step)

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py:
```diff
@@ -0,0 +1,381 @@
+import KratosMultiphysics
+import KratosMultiphysics.StructuralMechanicsApplication
+import math
+from structural_mechanics_analysis import StructuralMechanicsAnalysis
```
```suggestion
from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis
```

philbucher commented:
my main three points of critisism are that you should 
1. separate header-source
2. add a test (Otherwise no guarantees can be made ;) )
3. Still there are some debugging statements

Why would you return a value from the AnalysisStage? I thought this should be the main instance in a simulation?

RiccardoRossi commented:
regarding 1, ok.

regarding point 2 and 3 they go togheter: the goal of this analysis stage is to compute the ElasticityTensor which corresponds to the rve. 
As of now this is written to a file.

i was wondering if in the test i should write and read this file, or if i shall output it somehow. 

also i am not sure of where i should put the input parameters that belong to the "analysisstage"

RiccardoRossi committed:
splitting into header and cpp

philbucher commented:
So you would call this custom stage from the regular StructuralMechanicsAnalysis?

RiccardoRossi committed:
[skip CI] adding a test. test is still failing, however it detects a memory error to be investigates

loumalouomega committed:
Merge branch 'master' into RVE_periodic

RiccardoRossi commented on applications/StructuralMechanicsApplication/tests/rve_test/test_rve_analytic.py:
```diff
@@ -0,0 +1,57 @@
+        
+        model = KratosMultiphysics.Model()
+        simulation = RVEAnalysis.RVEAnalysis(model,parameters,boundary_mp_name,averaging_mp_name)
+        simulation.Run()
```
@loumalouomega this is the test that uncovers a memory bug (as well as a bug in the results) detected by valgrind. As of now it is skipped but it breaks deterministically 

@adityaghantasala most interestingly with the block builder and solver i get the exact prediction for this testcase. With this i mean that the AVERAGE stress is ok, nevertheless the solution in terms of displacements is not the one i would expect. This is actually a very strange result since it appears to be against the uniqueness of solution...
The point is that the expected result is that 

        displacement(nodeI) = strain_tensor * InitialPosition(nodeI)

this is periodic (with jump) and complies with the boundary conditions i am applying. Since the strain field is constant, stress is constant and like in a patch test FEM should deliver exact results.

From this analysis i get in average the correct stress, nevertheless the displacement field is not linear and hence the stress field is not constant. Periodicity condition IS respected

loumalouomega committed:
Cotire compilation issue

loumalouomega committed:
Include moved to cpp

loumalouomega committed:
[skip ci] Avoid error during debugging WIP

RiccardoRossi committed:
moving one function to the cpp

RiccardoRossi committed:
moving one function to the cpp

RiccardoRossi committed:
adding variable output-  to a dummy variable for the moment

RiccardoRossi committed:
stricted tolerance in test

RiccardoRossi committed:
Merge branch 'RVE_periodic' of https://github.com/KratosMultiphysics/Kratos into RVE_periodic

RiccardoRossi committed:
using by default the Block version appears to work

loumalouomega committed:
Minor clean up

loumalouomega committed:
[skip ci] More clean up

loumalouomega committed:
[skip ci] Adding some OMP

RiccardoRossi committed:
importing following new style

RiccardoRossi committed:
adding ELASTICITY_TENSOR variable

RiccardoRossi committed:
polishing

RiccardoRossi committed:
improving the test - now passing with blockbuilderandsolver but not with eliminationbuilderandsolver

RiccardoRossi committed:
cleaning up on exit

RiccardoRossi committed:
removing commented code

RiccardoRossi committed:
improving formatting some

RiccardoRossi commented:
@philbucher i think i addressed all of your concerns (and indeed code is much more polished than it used to be)

I am not sure of what you are asking about the StructuralAnalysisStage.

The RVEAnalysis is derived from that ... but it performs a completely different operation (Actually it applies 6 different periodicity conditons and uses that to estimate the ELASTICITY_TENSOR of an RVE).
This however also implies that one needs to construct explicitly from the main the RVEAnalysis.
Incidentally this highlights a limitation of the current mecanism:

I believe that our project parameters should eventually have something like

       stages_list =[
           {
                "stage_type" : StructuralAnalysisStage,
                "stage_parameters" : ...
           },
           {
                "stage_type" : RVEAnalysisStage,
                "stage_parameters" : ...
           }

of course this is not to be addressed in this PR, i am just telling it since if we already had this mechanism than i would be able to use the same Main as always

RiccardoRossi committed:
passing autopep8

RiccardoRossi commented:
@loumalouomega the new test is in a subdirectory. i am not clear on how to include it in the test suite...could you guide me?

loumalouomega commented:
> @loumalouomega the new test is in a subdirectory. i am not clear on how to include it in the test suite...could you guide me?

I can, let me go to your branch

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Updating test

loumalouomega committed:
Adding test to suite

loumalouomega commented:
@RiccardoRossi I added the test to the suite, I will continue debugging

loumalouomega committed:
MIssing KRATOS_API

loumalouomega committed:
Minor clean up

loumalouomega committed:
Ucomment

RiccardoRossi committed:
minor change to allow output

RiccardoRossi commented:
@philbucher i get a warning of "old format" been used, but i cannot see what is wrong.

would u mind taking a look?

philbucher commented:
> @philbucher i get a warning of "old format" been used, but i cannot see what is wrong.
> 
> would u mind taking a look?

@RiccardoRossi to me your input looks correct/up-to-date
can you please post the entire warning?

RiccardoRossi commented:
here goes the output

~~~
 |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ___ |  /           |
 ' /   __| _` | __|  _ \   __|
 . \  |   (   | |   (   |\__ \
_|\_\_|  \__,_|\__|\___/ ____/
           Multi-Physics 6.0.0-693bc1d484-FullDebug
Importing  KratosStructuralMechanicsApplication 
    KRATOS   ___|  |                   |                   |
           \___ \  __|  __| |   |  __| __| |   |  __| _` | |
                 | |   |    |   | (    |   |   | |   (   | |
           _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
Initializing KratosStructuralMechanicsApplication...
StructuralMechanicsAnalysis: Using the old way to create the processes, this will be removed!
Refer to "https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage" for a description of the new format 
_/
           Multi-Physics 6.0.0-693bc1d484-FullDebug
Importing  KratosStructuralMechanicsApplication 
    KRATOS   ___|  |                   |                   |
           \___ \  __|  __| |   |  __| __| |   |  __| _` | |
                 | |   |    |   | (    |   |   | |   (   | |
           _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
Initializing KratosStructuralMechanicsApplication...
StructuralMechanicsAnalysis: Using the old way to create the processes, this will be removed!
Refer to "https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage" for a description of the new format 
~~~

philbucher commented:
your processes are not wrapped in a "processes" block
=> check out https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users#analysisstage-usage gor how to use it (it is very simple to fix)

RiccardoRossi commented:
actually i have no processes at all...

RiccardoRossi commented:
I should have written: i deliberately have no processes (everything is done automatically)

philbucher commented:
ok now I understand

Fix incoming ...

RiccardoRossi committed:
minimal changes

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Updating B&S

RiccardoRossi committed:
Merge branch 'RVE_periodic' of https://github.com/KratosMultiphysics/Kratos into RVE_periodic

loumalouomega committed:
Reference solution

loumalouomega committed:
Reverting

RiccardoRossi committed:
Merge branch 'RVE_periodic' of https://github.com/KratosMultiphysics/Kratos into RVE_periodic

loumalouomega committed:
Clean up analysis

loumalouomega committed:
Initialize master/slave relation

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Adding check

loumalouomega committed:
Some corrections

loumalouomega committed:
More corrections

loumalouomega committed:
Removing duplicated computation

loumalouomega committed:
Correction

loumalouomega committed:
One more change

loumalouomega committed:
More changes

loumalouomega committed:
Correction

RiccardoRossi committed:
committing a working version of the block builder and solver - NOT TO MERGE - NEEDS A LOT OF POLISHING

loumalouomega committed:
Minor

loumalouomega committed:
DoF fix

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Last fix

RiccardoRossi committed:
omp fix

RiccardoRossi committed:
increasing accuracy

RiccardoRossi committed:
always resetting to the initial position - also displacement is initialized to zero for check

RiccardoRossi committed:
printing results

RiccardoRossi committed:
Merge branch 'RVE_periodic' of https://github.com/KratosMultiphysics/Kratos into RVE_periodic

RiccardoRossi committed:
Merge branch 'RVE_periodic' of https://github.com/KratosMultiphysics/Kratos into RVE_periodic

RiccardoRossi committed:
performance improvement

RiccardoRossi committed:
using the bounding box volume in computing the average

RiccardoRossi committed:
reducing tolerance in test to have it to pass also with the elimination BS

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py:
```diff
@@ -0,0 +1,353 @@
+# makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
this file should be names `rve_analysis.py`

RiccardoRossi replied:
ok

philbucher commented on applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:
```diff
@@ -0,0 +1,76 @@
+import KratosMultiphysics.StructuralMechanicsApplication as StructuralMechanicsApplication
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+import KratosMultiphysics.kratos_utilities as kratos_utilities
+import RVEAnalysis
```
```suggestion
from KratosMultiphysics.StructuralMechanicsApplication.rve_analysis import RveAnalysis
```

RiccardoRossi replied:
done (still calling it RVEAnalysis instead of RveAnalysis)

philbucher replied:
fine for me, as you explained me why :+1:

philbucher commented on applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:
```diff
@@ -0,0 +1,76 @@
+                self.assertAlmostEqual(
+                    abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
+
+        if not parameters["rve_settings"]["print_rve_post"].GetBool():
```
this check is not strictly needed, would also work without

RiccardoRossi replied:
i leave the check since i wish to be able to actually see the output if i set the print_rve_post to true

philbucher replied:
ok I see

I still have a suggestion for you: From what I understand you always print the output, even if you don't want to see it (in 99% of the cases this test is run)
=> this is time consuming
=> I recommend to do it the other way round, explicitly only adding the printing of output if you want it, take a look what I am doing [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MeshMovingApplication/tests/test_ale_fluid_solver.py#L48-L101). This way the test won't get slowed down for nothing

RiccardoRossi replied:
It is only been printed if the parameters["rve_settings"]["print_rve_post"] is True

philbucher replied:
but why would you then need to delete the files at all? If you want to print them you would want to look at them right?

philbucher replied:
ok I got it now, fixing it

philbucher commented on applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:
```diff
@@ -0,0 +1,76 @@
+    def _aux_rve_computation(self, parameters):
+
+        model = KratosMultiphysics.Model()
+        simulation = RVEAnalysis.RVEAnalysis(model, parameters)
```
```suggestion
        simulation = RveAnalysis(model, parameters)
```

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/RVEAnalysis.py:
```diff
@@ -0,0 +1,353 @@
+from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis
+
+
+class RVEAnalysis(StructuralMechanicsAnalysis):
```
```suggestion
class RveAnalysis(StructuralMechanicsAnalysis):
```

RiccardoRossi replied:
I would prefer keeping the capitals here, as it is 

Represenative Volume Element

philbucher replied:
ok, no problem

philbucher reviewed (commented):
some comments related to naming and importing

roigcarlo committed:
Missing include

loumalouomega committed:
Merge remote-tracking branch 'origin/RVE_periodic' into RVE_periodic

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Commenting post-process of RVE test

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Correction on relative path

loumalouomega committed:
Actual relative path fix

philbucher commented on applications/StructuralMechanicsApplication/tests/test_rve_analytic.py:
```diff
@@ -0,0 +1,95 @@
+
+import os
+
+class controlledExecutionScope:
```
please use the new class in `kratos_utilities` for this

RiccardoRossi replied:
@philbucher i am not clear what you are asking for. could u point me to an example?

philbucher replied:
https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/tests/structural_mechanics_test_factory.py#L13

RiccardoRossi replied:
thx, i think i corrected this

philbucher reviewed (changes requested):
some things need to be cleaned up (my comments and the changes to the B&S)

RiccardoRossi commented:
ok. Anyway this goes in after release

loumalouomega commented:
> ok. Anyway this goes in after release

@RiccardoRossi can you approve the PR for my B&S?

RiccardoRossi committed:
reverting version from core

RiccardoRossi committed:
merging master into the branch

RiccardoRossi committed:
committing name change

RiccardoRossi committed:
minor change in the importing mechanism

RiccardoRossi committed:
name change and reverting to master version of builder and solvers

RiccardoRossi committed:
changing scope function

philbucher commented:
@RiccardoRossi if you are done with the changes I can also do some minor cleanup myself, this would be more efficient than having to iterate here :)

Just give me a go and I will do it according to the latest standards

RiccardoRossi commented:
@philbucher ok thx.

RiccardoRossi commented:
mmm @philbucher i mean, i just commiteed my latest change

philbucher commented:
I cleaned up a bit, but I noticed that there are some occasional errors (even without my modifications, thats why I haven't pushed them yet):
Can you take a look please?
~~~
======================================================================
FAIL: test_rve_computation_elimination_version (__main__.TestRVESimplestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_rve_analytic.py", line 36, in test_rve_computation_elimination_version
    self._aux_rve_computation(parameters)
  File "test_rve_analytic.py", line 74, in _aux_rve_computation
    abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
AssertionError: nan != 0.0 within 5 places

----------------------------------------------------------------------
~~~

RiccardoRossi committed:
Merge branch 'master' into RVE_periodic

RiccardoRossi committed:
desactivating the use of the elimination builder and solver

RiccardoRossi commented:
mmm i tried with valgrind and it looks like there is still something wrong with the EliminationB&S

@loumalouomega i am afraid this is on you.If you prefer i can take care of it

Right now i commented the test with the elimination builder and solver and left the default (blockB&S)
@philbucher could u see if it works correctly for you?

loumalouomega commented:
> mmm i tried with valgrind and it looks like there is still something wrong with the EliminationB&S
> 
> @loumalouomega i am afraid this is on you.If you prefer i can take care of it
> 
> Right now i commented the test with the elimination builder and solver and left the default (blockB&S)
> @philbucher could u see if it works correctly for you?

When you say elimination you mean the elimination with constraints?

philbucher commented:
The problems I had were only with the elimination-based

loumalouomega committed:
Merge branch 'master' into RVE_periodic

RiccardoRossi commented:
@loumalouomega yes. In any case i would still merge this branch if it works, since the fix is unrelated

loumalouomega commented:
> @loumalouomega yes. In any case i would still merge this branch if it works, since the fix is unrelated

In my machine the B&S works. Which is the problem?

loumalouomega commented:
Besides, the test is still generating the auxiliar txt. This should not happen

RiccardoRossi commented:
@loumalouomega  the txt is not auxiliar, it is the output of the RVEanalysis

regarding the Elimination B&S just try to compile this branch with FullDebug & -D_GLIBCXX_DEBUG and run it with valgrind

philbucher commented:
@RiccardoRossi I will push the cleanup on Monday latest then this can go in, since the problems are in the B&S
Do you agree?

loumalouomega committed:
Reactivating test

loumalouomega committed:
[Core] Several fixes on ResidualBasedEliminationBuilderAndSolverWithConstraints

This solves problems from valgrind and checks if the constraints area active (this was not taken into account until now, added for consistency with the block B&S)

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Revert "Removing evil pointer on member variables"

This reverts commit 43cd9f8c719d33932a2311fd09633a2f40b5865e.

loumalouomega committed:
Removing evil pointer on member variables

loumalouomega commented:
> @loumalouomega the txt is not auxiliar, it is the output of the RVEanalysis
> 
> regarding the Elimination B&S just try to compile this branch with FullDebug & -D_GLIBCXX_DEBUG and run it with valgrind

Yes, already done. I created an independent PR on  #4400

loumalouomega committed:
Merge branch 'master' into RVE_periodic

loumalouomega committed:
Minor

RiccardoRossi commented:
@philbucher ok thx

philbucher committed:
remove unused import

philbucher committed:
adding name of mainmdelpart

philbucher committed:
remove comment

philbucher committed:
another unused import

philbucher committed:
using functions from ModelPart

philbucher committed:
using function from variableutils

philbucher committed:
optionally adding output

philbucher committed:
minor cleanup in header

philbucher committed:
minor cleanup on cpp

philbucher commented:
I pushed some cleanup, hope its ok

Tests still running, but I sometimes still get the following error (FullDebug with two threads):
~~~
======================================================================
FAIL: test_rve_computation_elimination_version (__main__.TestRVESimplestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_rve_analytic.py", line 33, in test_rve_computation_elimination_version
    self._aux_rve_computation(parameters)
  File "test_rve_analytic.py", line 103, in _aux_rve_computation
    abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
AssertionError: nan != 0.0 within 5 places
~~~

loumalouomega commented:
> I pushed some cleanup, hope its ok
> 
> Tests still running, but I sometimes still get the following error (FullDebug with two threads):
> 
> ```
> ======================================================================
> FAIL: test_rve_computation_elimination_version (__main__.TestRVESimplestTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "test_rve_analytic.py", line 33, in test_rve_computation_elimination_version
>     self._aux_rve_computation(parameters)
>   File "test_rve_analytic.py", line 103, in _aux_rve_computation
>     abs(Cestimated[i, j] - Canalytic[i, j])/(l+2*G), 0.0, 5)
> AssertionError: nan != 0.0 within 5 places
> ```

Why it only fails in others computer?, well take into account I reverted the pointers due to @RiccardoRossi suggestion (and this was one of the reasons of valgrind messages)

philbucher commented:
I had the latest modifications

loumalouomega commented:
> I had the latest modifications

Yes, but I reverted the pointers change due to @RiccardoRossi suggestions (in any case I cannot duplicate the error)

loumalouomega committed:
Merge branch 'master' into RVE_periodic

RiccardoRossi commented:
@loumalouomega but did u fix the error in some other way?

loumalouomega commented:
> @loumalouomega but did u fix the error in some other way?

YEs, thta's why I did it. I assume that the error must be in some swap...but I cannot replicate it

loumalouomega committed:
Commenting until fixed random fail on elimination B&S

loumalouomega reviewed (approved):
Approving and commenting test until fixed random fail on elimination B&S

loumalouomega commented:
@RiccardoRossi feel free to merge
